### ISSUE TITLE
Accessing values of environment variables

### ISSUE THREAD

**Issue by newhoggy** [Gabriel439/Haskell-Dhall-Library#23]
Is there a way to evaluate an environment variable in a dhall expression?

**Comment by Gabriel439** [comment]
No.  Dhall can only import values via file paths or URLs

**Comment by newhoggy** [comment]
Do you see handling of environment variables as something that should be external do Dhall?

**Comment by Gabriel439** [comment]
Actually, I could add support for reading in Dhall expressions via environment variables, using something like this syntax:

```
{ foo = ${BAR}
}
```

However, there are two issues with that:

*   I might want to reserve the same syntax for interpolating `Text` into a string literal, like this:

    ```
    let txt = "Hello" in "${txt} world!"
    ```

    ... and it might be confusing to use the same syntax in two different ways

*   Imported environment variables would have to be valid Dhall expressions, meaning that if you
    wanted to import `Text`, then you'd have to define the environment variable like this:

    ```
    EXAMPLE='"Test"'
    ```

**Comment by newhoggy** [comment]
I can see your point.

What if there was an implicit global `let` binding that has all the environment variables in them as strings?

For example:

```
{ foo = ENV.BAR
}
```

would imply:

```
let ENV = { BAR = <string-value-of-BAR> } in
{ foo = ENV.BAR
}
```

**Comment by Gabriel439** [comment]
What is the motivation for providing support for environment variables, though?  The reason I ask is that any program that takes environment variables implicitly could be done by reformulating the program as a function.  For example, your code could be written as:

```haskell
    λ(ENV : { FOO : Text, BAR : Text })
→   { foo = ENV.BAR }
```

... and instead of running:

```bash
$ FOO=foo BAR=bar dhall <<< './example'
```

... you could supply the desired variables as an ordinary function argument when invoking the expression:

```bash
$ dhall <<< './example { FOO = "foo", BAR = "bar" }'
```

... or if you are really constrained to obtain the values from environment variables, you could even do:

```bash
$ dhall <<< "./example { FOO = \"${FOO}\", BAR = \"${BAR}\" }"
```

**Comment by newhoggy** [comment]
In the last case, it will mostly work except when the environment variable needs to be escaped, which is the tricky bit.

Perhaps all we need is a `env-to-dhall` command line tool that does all the necessary escaping?

```
dhall <<< "λ(ENV : $(env-to-dhall)) → { foo = ENV.BAR }"
```

Or:

```
dhall <<< "λ(ENV : $(env-to-dhall)) → ./config"
```

**Comment by Gabriel439** [comment]
Hmmm.  I would prefer reading environment variables directly within Dhall as a built-in mechanism rather than provide an `env-to-dhall` executable

The main issue with `env-to-dhall` or an `Env` record in general is that `dhall` does not allow records to hold fields which are types or kinds.  So if one of your environment variables is a type or kind you will get a type error when type-checking the `Env` record

The second issue (when I thought about this more) is that an `env-to-dhall` mechanism does not support transitively resolving environment variables.  In other words, if an environment variable references another environment variable then Dhall should support that (and do basic cycle checking like it does for files/URLs).

So what I'll probably do is add something like `env://FOO` to reference the environment variable `FOO` in Dhall.  However, `FOO` will need to be a valid Dhall expression, meaning that if you want to import a `Text` string then the contents of `FOO` need to be quoted.

So in other words, you could do:

```bash
$ FOO='"foo"' BAR='1' dhall <<< '{ foo = env://FOO , bar = env://BAR }'
{ bar : Integer, foo : Text }

{ bar = 1, foo = "foo" }
```

How does that sound?

**Comment by newhoggy** [comment]
That sounds really good.

Although being able embed variables as string is also useful.  I don't mind having to type extra to get it.  For example, I'd be happy with `env+text://FOO`.  Even `file+text://pubrung.gpg` would be helpful too.

**Comment by Gabriel439** [comment]
I just realized that `env://FOO` is a misleading URI since it implies that `FOO` is the authority as opposed to the path.  It seems like `env:FOO` is the more appropriate URI

I also like the idea of using `+text` in the scheme to denote that we expect unquoted `Text`.  I'll implement that, too

**Comment by newhoggy** [comment]
Good spot.  Thanks!

**Comment by Gabriel439** [comment]
Alright, the first part of this is done in 11ceab1dfeb9ed9a25dab717b4fe24ffaf7d320e (supporting `env:NAME` syntax).  The `env+text` and `file+text` support will come in a subsequent commit

**Comment by newhoggy** [comment]
Looking good!

**Comment by newhoggy** [comment]
I just discovered a case that might need better `dhall` support:

```
circle-cli-exe:
↳ ./project.config

Error: Invalid input

project.config:1:17: error: expected: "''",
    "(", "+", "-", "[", "\8704",
    "\955", "\\", "forall", "if",
    "let", "merge", built-in value,
    double, import, integer, label,
    list literal, natural,
    record literal, record type,
    string, union literal,
    union type
{ environment = ~/.circle/project-config/haskell-works.config
                ^
```

**Comment by newhoggy** [comment]
At the moment, I'm required to put the absolute path in:

```
{ environment = ~/.circle/project-config/haskell-works.config
}
```

**Comment by Gabriel439** [comment]
Could I propose an alternative syntax for importing raw `Text`?  I was thinking of adding an `as Text` suffix to the path in order to import it as a raw `Text` literal, like this:

```haskell
{ foo = /foo as Text
, bar = env:HOME as Text
, baz = http://example.com as Text
}
```

Would that be alright?

**Comment by Gabriel439** [comment]
I put up a PR for the `as Text` proposal here: https://github.com/Gabriel439/Haskell-Dhall-Library/pull/34

**Comment by newhoggy** [comment]
👍

**Comment by newhoggy** [comment]
Thanks so much!

**Comment by Gabriel439** [comment]
You're welcome! :)

**Comment by feliksik** [comment]
This is actually a very nice feature to manage global cross cutting concerns nicely. This makes it very powerful, especially since it is also evaluating file imports. Great job!

**Comment by Gabriel439** [comment]
@feliksik: Thanks 🙂
